### DISCUSSION TITLE
Run in Cloudflare Worker + Durable Object

### DISCUSSION THREAD

**Discussion by jokester** [socketio/socket.io#5019] [Ideas]
I am interested by serverless websocket server and cost saving (via [WebSocket Hibernation](https://developers.cloudflare.com/durable-objects/examples/websocket-hibernation-server/)) CF provide, so I made a [PoC](https://sio-worker-demo.ihate.work/) / [code](https://github.com/jokester/limb).

Many pieces are lacking including persisted state, WS Hibernation, clustering or load balancing but it basically works.

I don't have a clear roadmap though, so ideas welcome 🤔 

======

Update: a working npm package at https://github.com/jokester/socket.io-serverless / https://www.npmjs.com/package/socket.io-serverless .

**Comment by jokester** [comment]
It won't be easy to make this an official feature of socket.io. To build and run it in CF worker it required quite a few tweaks in `socket.io` `engine.io` and other packages.

If a reusable artifact interests other people I guess the easier way is to create a custom `socket.io` JS bundle.

**Comment by darrachequesne** [comment]
@jokester I guess what we could do is create an engine based on the Cloudflare WebSocket server, and then bind it to the high-level Socket.IO server.

The engine manages the low-level connections (HTTP long-polling or WebSocket). The `engine.io` package is one implementation of an engine, based on the Node.js HTTP server: https://github.com/socketio/engine.io

**Reply by jokester** [reply]
Yes. CF has its own WebSocket impl so code relied on `ws` or other WS server is not used.

My PoC basically fused CF WebSocket to `eio.Socket` `eio.WebSocket` (I only wanted to support websocket transportation) to reuse their code. Above that `sio.{Server,Client,Namespace}` mostly run likely they do in Node.js 👍🏽

**Comment by jokester** [comment]
Months later: I think I have a working version now: https://github.com/jokester/socket.io-serverless .

It is based on engine.io (WebSocketTransport) and socket.io (Server / Client / Socket / Namespace / Adapter) code, but rewires a lot to work in Cloudflare Worker , especially to support hibernation / alarm.

I learned a lot making this lol.

**Comment by petrbela** [comment]
Check out https://www.partykit.io/. It's a WebSocket server built on top of CF/DurableObjects with a TypeScript SDK.
